Refactor front-end for new app stack

br3ndonland opened the pull request:
# Refactor front-end for new app stack

## Updates

This collection of commits improves the front-end functionality and user experience.

- **General**
  - Refactor with JavaScript ES6+ syntax
- **Geolocation**
  - Geolocate user with success and failure callbacks as before
    - Note that I'm currently not seeing an error message when the user denies the geolocation prompt, even though I have the failure callback function.
  - Separate geolocation success from generation of map, markers and table
    - Previously, we had the map and markers generated as part of the `success()` function. This limits the ability to receive coordinates from other methods, like text search or other lat/lng JavaScript objects.
    - I separated these functions for flexibility, and linked them with callbacks. Successful geolocation returns a `Position` object that contains `coords.latitude` and `coords.longitude`. I store the latitude and longitude in a `location` object, which is passed into the `showStops` function, and then call the function with the object using `showStops(location)`. If I add text search or university search later, I can generate another `location` object, and run `showStops(location)` to do the same thing.
- **Map**
  - Retain same center and zoom as before
  - Eliminate call to Google Maps Places API to minimize API calls and improve speed
  - Scope all JavaScript under `initMap()` to create closure and run entire file when map loads
  - Wrap the contents of `initMap()` in a try/catch block for general error handling
  - Add a map to the universities page
- **Stops and markers**
  - Update *app.py* to only return stops with `wheelchair_boarding: 1`
  - Drop map markers for each location
    - *index.html*: One map marker for user location, with star label and infoWindow. Additional markers for wheelchair accessible stops.
    - *universities.html*: Markers for university locations.
  - Create infoWindow for each marker with link to Google Maps directions
  - Fit map bounds to markers
  - Resize map automatically when screen resizes
- **Tables**
  - Generate tables with vanilla JavaScript and Bootstrap classes instead of jQuery DataTables plugin
    - The new tables aren't currently sortable, but we can consider sorting in the future.
    - We should also consider incorporating the university data into our database instead of storing as JSON.
- **Header**
  - Highlight nav link for current page by adding `active` class with JavaScript
  - Update header icon
- **Footer**
  - Move footer date calculation to external JavaScript file *date.js*
  - Calculate footer date with vanilla JavaScript instead of jQuery
  - Adjust HTML and CSS to prevent map and tables from running over footer
- **About page**
  - Add a page at */about* to explain the app
- **Service Worker**
  - Update service worker with new assets

## Notes

Note that the search box is not yet operational. I plan to implement Elasticsearch instead of the Google Places API.

I wrote the new JavaScript in [Standard Style](https://standardjs.com/) (without unnecessary semicolons). I prefer Standard Style, and hopefully it's okay with you. If it's a problem I can autoformat in the semicolons ("semistandard" style as we had before).

AbdouSeck commented:
@br3ndonland, this is fantastic! The looks and everything! I like seeing the markers bouncing and enjoyed seeing the one marker that marks exactly (with **You**) where I am. The icon is a great choice.
I also like that only stops with wheelchair access are returned by the API. This actually makes a whole lot more sense.

I have 2 questions and 1 suggestion.

Questions:

1. Is there any reason why the location search feature is no longer available? The application seems to only show me the stops based on the geolocation Position it was able to gather from me.

2. Since I am unable to fully use the search feature, I could not test what would happen if the API returned no stop. Do you know how this is handled? No markers on the map or some kind of message indicating that there no stops? Or both?

Suggestion:

Can you please set the `target` attribute of the anchor (`<a></a>`) tags to some `"_blank"`? Clicking on most links (markers and table rows) opens in the current tab.

`jshint` freaked out when I opened `js` files because all the semi-colons are gone. This decision definitely threw me off a little bit. But like I usually say, as long as compile errors aren't encountered and the code is easy to follow (which it is truly is), I am fine with it.

Sorry about the delay in getting this review in. I will try and stay on top of it.

br3ndonland commented:
Thanks for your review @AbdouSeck! I'm glad you like the updates. After we resolve this PR, I'll re-test the Docker build and update PR #46.

I added some commits to address your comments:

1. **Text search:** The text search will be the next feature. We previously used the Google Maps Places API to search for stops. This requires extra API calls, potentially slowing down the app. We also have our own database of stops now. We don't need to query Google if we only need data from MBTA. I plan to use [Elasticsearch](https://www.elastic.co/products/elasticsearch) (in a separate Docker container) to search through the stop names. Working on it now!
2. **Alerts:** Great point. We should provide user feedback when the app doesn't have geolocation or stop results. I added some JavaScript to drop an alert `<div>` above the map.
    - Alert if browser lacks geolocation support, though this is [unlikely](https://caniuse.com/#search=geolocation).
    - Alert when user denies geolocation. The alert mirrors the console log message.
    - Alert when no stops are returned. Try replacing the geolocation coordinates with this:

      ```js
      // Within the `success` function:
      let location = {
        lat: 420.35947,
        lng: -710.09296,
        name: 'You'
      }
      console.log(`Testing an erroneous location: `, position)
      ```

      These are the same error coordinates as the `test_stops` unit test in the *app_test.py* module.

      ![mbtaccess-pr-45-01](https://user-images.githubusercontent.com/26674818/44216722-9507d780-a143-11e8-9a82-609ddda211d6.png)

Your suggestion for `<a>` link elements is great. I added `target="_blank"` to external hyperlinks throughout the site.

AbdouSeck commented:
@br3ndonland, I like the idea behind using `ElasticSearch` but the reason we were using the Place API was to actually get the geolocation of any arbitrary place the end user decides to input. I don't think we will be able to do that with ElasticSearch. For instance, if I wanted to know how many wheelchair accessible stops are near `77 Mass Ave, Cambridge` while I am at home in `Malden`, the `ElasticSearch` feature will have no way of knowing which stops to suggest while I am typing `77 Mass Ave, Cambridge...`. Or does it?

Also, I think the anchor tags for the markers are still not set to `_blank`. But I am going to merge this pull request because that change can be made later, or not. Thanks for the work on this. It's great stuff!
